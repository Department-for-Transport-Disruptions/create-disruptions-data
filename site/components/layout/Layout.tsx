import Head from "next/head";
import Link from "next/link";
import { useRouter } from "next/router";
import { PropsWithChildren, ReactElement, useEffect, useState } from "react";
import { createPortal } from "react-dom";
import { ErrorInfo } from "../../interfaces";
import { buildTitle } from "../../utils";
import CookieBanner from "./CookieBanner";
import Footer from "./Footer";
import PhaseBanner from "./PhaseBanner";

interface LayoutProps {
    title: string;
    description: string;
    errors?: ErrorInfo[];
    hideCookieBanner?: boolean;
    hideHelp?: boolean;
    disableBackButton?: boolean;
}

const Help = (): ReactElement => (
    <div>
        <h2 className="govuk-heading-s">Help and Support</h2>
        <p className="govuk-body">
            If you are having problems, please contact the Create Disruption Service via this link:{" "}
            <Link href="/contact" className="govuk-link govuk-!-font-size-19">
                Contact us
            </Link>
        </p>
    </div>
);

export const BaseLayout = ({
    title,
    description,
    errors = [],
    hideCookieBanner,
    children,
    hideHelp,
    disableBackButton = false,
}: PropsWithChildren<LayoutProps>): ReactElement => {
    const [showBanner, setShowBanner] = useState(false);

    useEffect(() => {
        setShowBanner(true);
    }, [setShowBanner]);

    let element = null;
    if (typeof document !== "undefined") {
        element = document.getElementById("js-cookie-banner");
    }

    const router = useRouter();
    return (
        <>
            <Head>
                <link rel="icon" href="/assets/rebrand/images/favicon.ico" />
                <title>{buildTitle(errors, title || "Create Disruptions")}</title>
                <meta name="description" content={description || "Create Disruptions"} />
                <meta name="viewport" content="width=device-width, initial-scale=1" />
                <meta charSet="utf-8" />
            </Head>

            {!hideCookieBanner && showBanner && element && <div>{createPortal(<CookieBanner />, element)}</div>}
            <PhaseBanner />

            <div className="govuk-width-container">
                {!disableBackButton && (
                    <button type="button" onClick={() => router.back()} className="govuk-back-link mb-7">
                        Back
                    </button>
                )}
                <main className="govuk-main-wrapper" id="main-content">
                    {children}
                </main>
                {!hideHelp && <Help />}
            </div>
            <Footer />
        </>
    );
};

export const FullColumnLayout = ({
    title,
    description,
    errors = [],
    children,
    hideCookieBanner = false,
    hideHelp = false,
    disableBackButton = false,
}: PropsWithChildren<LayoutProps>): ReactElement => (
    <BaseLayout
        title={title}
        description={description}
        errors={errors}
        hideCookieBanner={hideCookieBanner}
        hideHelp={hideHelp}
        disableBackButton={disableBackButton}
    >
        <div className="govuk-grid-row">
            <div className="govuk-grid-column-full">{children}</div>
        </div>
    </BaseLayout>
);

export const TwoThirdsLayout = ({
    title,
    description,
    errors = [],
    children,
    hideCookieBanner = false,
    hideHelp = false,
    disableBackButton = false,
}: PropsWithChildren<LayoutProps>): ReactElement => (
    <BaseLayout
        title={title}
        description={description}
        errors={errors}
        hideCookieBanner={hideCookieBanner}
        hideHelp={hideHelp}
        disableBackButton={disableBackButton}
    >
        <div className="govuk-grid-row">
            <div className="govuk-grid-column-two-thirds">{children}</div>
        </div>
    </BaseLayout>
);

export default TwoThirdsLayout;
