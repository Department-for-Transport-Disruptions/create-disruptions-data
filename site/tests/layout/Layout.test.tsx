import * as React from 'react';
import renderer from 'react-test-renderer';
import Layout from '../../components/layout/Layout';
import {describe, it, expect} from 'vitest'

describe('Layout', () => {
    it('should render correctly', () => {
        const tree = renderer.create(<Layout title="title" description="description" />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
