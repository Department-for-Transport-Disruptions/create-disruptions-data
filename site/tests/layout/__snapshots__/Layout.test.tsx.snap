// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Layout > should render correctly 1`] = `
[
  <div
    className="app-phase-banner__wrapper"
  >
    <div
      className="govuk-phase-banner max-w-[960px] w-[90%] border-0 mx-auto app-phase-banner app-width-container"
    >
      <p
        className="govuk-phase-banner__content"
      >
        <strong
          className="govuk-tag govuk-phase-banner__content__tag"
        >
          alpha
        </strong>
        <span
          className="govuk-phase-banner__text"
        >
          This is a new service â€“ your
           
          <a
            className="govuk-link"
            href="/feedback"
            id="feedback-link"
          >
            feedback
          </a>
           
          will help us to improve it.
        </span>
      </p>
    </div>
  </div>,
  <div
    className="govuk-width-container"
  >
    <main
      className="govuk-main-wrapper"
    >
      <div
        className="govuk-grid-row"
      >
        <div
          className="govuk-grid-column-two-thirds"
        />
      </div>
    </main>
    <div>
      <h2
        className="govuk-heading-s"
      >
        Help and Support
      </h2>
      <p
        className="govuk-body"
      >
        If you are having problems, please contact the Create Disruption Service via this link:
         
        <a
          className="govuk-link govuk-!-font-size-19"
          href="https://www.bus-data.dft.gov.uk/contact/"
        >
          Contact us
        </a>
      </p>
    </div>
  </div>,
  <footer
    className="govuk-footer "
    role="contentinfo"
  >
    <div
      className="govuk-width-container "
    >
      <div
        className="govuk-footer__meta"
      >
        <div
          className="govuk-footer__meta-item govuk-footer__meta-item--grow"
        >
          <h2
            className="govuk-visually-hidden"
          >
            Support links
          </h2>
          <ul
            className="govuk-footer__inline-list"
          >
            <li
              className="govuk-footer__inline-list-item"
            >
              <a
                className="govuk-footer__link"
                href="/contact"
              >
                Help
              </a>
            </li>
            <li
              className="govuk-footer__inline-list-item"
            >
              <a
                className="govuk-footer__link"
                href="/cookies"
              >
                Cookies
              </a>
            </li>
            <li
              className="govuk-footer__inline-list-item"
            >
              <a
                className="govuk-footer__link"
                href="/contact"
              >
                Contact
              </a>
            </li>
            <li
              className="govuk-footer__inline-list-item"
            >
              <a
                className="govuk-footer__link"
                href="/accessibility"
              >
                Accessibility
              </a>
            </li>
            <li
              className="govuk-footer__inline-list-item"
            >
              <a
                className="govuk-footer__link"
                href="https://www.gov.uk/help/terms-conditions"
              >
                Terms and conditions
              </a>
            </li>
            <li
              className="govuk-footer__inline-list-item"
            >
              <a
                className="govuk-footer__link"
                href="/privacy"
              >
                Privacy
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </footer>,
]
`;
